Add tests for "Get Title" command

jugglinmike opened the pull request:
@andreastt this is not yet ready for `master` because it relies on behavior in
`wdclient` that is both undocumented and non-spec-compliant.  Specifically: the
setting of the `unhandledPromptBehavior` capability. Although I've been
validating my work using geckodriver and Firefox exclusively, support for the
new capabilities format has only just landed and is not yet released.  That
makes the tests concerning `unhandledPromptBehavior` especially prone to
programming error on my part.

Still, I thought it would be good to submit my work now that I have rebased it
on top of gh-4756. This will let us start talking about it earlier and
hopefully result in a smoother review process over all.

If I'm reading the `wdclient` source code and the specification correctly, the
current format is:

```json
{
  "requiredCapabilities": {},
  "desiredCapabilities": {}
}
```

Where the new format is:

```json
{
  "capabilities": {
    "alwaysMatch": {},
    "firstMatch": {}
  }
}
```

I'm not sure how you would like to handle this, but I have a suggestion. How
about we update the "create session" logic to use the new "capabilities" format
and then extend the `Session#start` method to accept capabilities as an
optional parameter?

<!-- Reviewable:start -->

<!-- Reviewable:end -->

jugglinmike committed:
[webdriver] Add tests for "Get Title" command

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/216969620)

# Firefox (nightly channel) #
Testing web-platform-tests at revision cdf5906f3add15a4eb4dfb99af1f5c1eb35fe273
Using browser at version BuildID 20170330100130; SourceStamp 60d7a0496a3673450ddbc37ec387525148c32604
Starting 10 test iterations
No tests run.

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/216969621)

# Chrome (unstable channel) #
Testing web-platform-tests at revision cdf5906f3add15a4eb4dfb99af1f5c1eb35fe273
Using browser at version 59.0.3053.3 dev
Starting 10 test iterations
No tests run.

andreastt commented:
@jugglinmike geckodriver HEAD supports the new-New Session negotiation. This will become available in 0.16. We should support `{capabilities: {alwaysMatch: …, firstMatch: …, …}}` in what’s available from HEAD post 0.15.

jugglinmike commented:
I'll see what I can do about building geckodriver from source.

andreastt commented on webdriver/get-title.py:
```diff
@@ -0,0 +1,268 @@
+# 3. Return success.
+def test_title_handle_prompt_dismiss(session, create_dialog):
+    document = inline("<title>This is the document's title</title>")
+    session.required_capabilites = {"unhandledPromptBehavior":"dismiss"}
```
Space after `:`.

andreastt replied:
Also modification of internal property, which you might want to change when you rebase.

jugglinmike replied:
Fixed

jugglinmike replied:
Will do. On that note: do you have commit rights for `wpt-tools`? https://github.com/w3c/wpt-tools/pull/198

andreastt replied:
I do, but I see someone merged that already.

andreastt commented on webdriver/get-title.py:
```diff
@@ -0,0 +1,268 @@
+    document = inline("<title>This is the document's title</title>")
+    session.required_capabilites = {"unhandledPromptBehavior":"dismiss"}
+    session.start()
+    session.send_session_command("POST", "url", dict(url=document))
```
You’re not testing the URL command here. I assume you could just use `session.url = document`.

jugglinmike replied:
Sounds good to me

andreastt commented on webdriver/get-title.py:
```diff
@@ -0,0 +1,268 @@
+    result = session.transport.send("GET",
+                                    "session/%s/title" % session.session_id)
+
+    assert_success(result, "This is the document's title</title>")
```
This string seems wrong. I don’t think `</title>` should be included.

jugglinmike replied:
I'm not sure how that crept in there. Anyway, fixed.

andreastt commented on webdriver/get-title.py:
```diff
@@ -0,0 +1,268 @@
+    assert_success(result, "This is the document's title</title>")
+    assert_alert_handled(None)
+
+    assert_confirm_handled = create_dialog("confirm")
```
I don’t understand the usage of these two asserts. Can we make it clearer?

andreastt commented on webdriver/get-title.py:
```diff
@@ -0,0 +1,268 @@
+    result = session.transport.send("GET",
+                                    "session/%s/title" % session.session_id)
+
+    assert_success(result, "This is the document's title</title>")
```
Same stray `</title>`.

andreastt commented on webdriver/get-title.py:
```diff
@@ -0,0 +1,268 @@
+    result = session.transport.send("GET",
+                                    "session/%s/title" % session.session_id)
+
+    assert_success(result, "This is the document's title</title>")
```
Stray `</title>`.

andreastt commented on webdriver/get-title.py:
```diff
@@ -0,0 +1,268 @@
+# 3. Return success.
+def test_title_handle_prompt_accept(session, create_dialog):
+    document = inline("<title>This is the document's title</title>")
+    session.required_capabilites = {"unhandledPromptBehavior":"accept"}
```
Space after `:`.

andreastt commented on webdriver/get-title.py:
```diff
@@ -0,0 +1,268 @@
+#
+# 3. Return success.
+def test_title_handle_prompt_accept(session, create_dialog):
+    document = inline("<title>This is the document's title</title>")
```
This same document is used throughout many tests of this file, and that is fine. But could we in that case please reduce the length of it?

jugglinmike replied:
Certainly. Going with "WD doc title" for now

andreastt commented on webdriver/get-title.py:
```diff
@@ -0,0 +1,268 @@
+
+    result = session.transport.send("GET",
+                                    "session/%s/title" % session.session_id)
+    assert_success(result, "Foobar")
```
Missing assertiont hat it matches `document.title`.

jugglinmike replied:
Can you elaborate? I'm interpreting your comment to mean an assertion like:

    assert client.execute_script('return document.title;') == 'Foobar'

...but that looks like a test for [the `title` attribute in HTML](https://html.spec.whatwg.org/#document.title) rather than a WebDriver test.

andreastt replied:
And testing a string literal (which by the way is fine) is also implicitly making assumptions about HTML through the power of WebDriver returning `document.title`.

The point is that we want to test that the returned string matches exactly what `document.title` gives us. Arguably this is more important in the test where you deal with escaping than here, though.

jugglinmike replied:
Oh, now I understand. This is really interesting. It's not so much that we want to test the value of `document.title` itself; it's that we want to test that the command's result is consistent. So even if the browser is completely wrong about `document.title`, it should still pass the WebDriver test for "Get Title" because the WebDriver specification is written in terms of that value.

It seems like I ought to re-write all of these assertions to be in terms of `session.execute_script('return document.title')`. Does that seem overly pedantic to you?

andreastt replied:
I think it makes the most sense to call `session.execute_script("return document.title")` in the test you have for escape values, but if you feel like it I guess you can add a `document_title` fixture returning that or something, and make more use of it.

Anyway, I don’t mean to be overly pedantic about this, so it’s fine if you just call that once. We can make incremental improvements later as well.

jugglinmike replied:
I don't think it's overly pedantic at all, and I didn't mean to suggest
otherwise (that would be the pot calling the kettle black). Sorry for the
miscommunication.

andreastt commented on webdriver/get-title.py:
```diff
@@ -0,0 +1,268 @@
+# 4. Return success with data title.
+def test_title_after_modification(session):
+    session.url = inline("<title>Initial</title><h2>Hello</h2>")
+    session.execute_script("document.title = 'updated';")
```
Superfluous `;`.

andreastt commented on webdriver/get-title.py:
```diff
@@ -0,0 +1,268 @@
+    result = session.transport.send("GET",
+                                    "session/%s/title" % session.session_id)
+
+    assert_success(result, "a b c d e")
```
In this case I would check it against `document.title` instead of a string expectation.

jugglinmike replied:
Can you say why? At first glance, using `document.title` for this would seem to introduce a dependency on functionality that is not directly related to the behavior under test. E.g. if `document.title` were to return a different result, then we would want to write a test in HTML for that.

andreastt replied:
See reply above.

andreastt commented on webdriver/support/fixtures.py:
```diff
@@ -115,3 +117,83 @@ def inner(path, query="", fragment=""):
+
+        session.send_session_command("POST",
+                                     "execute/sync",
+                                     dict(script="window.open();", args=[]))
```
Why are you using `session.send_session_command` here when you can use `session.execute`?

andreastt replied:
Opening a new window could be done in a separate fixture, returning the window handle of the new window.

jugglinmike replied:
I'm happy to use the higher-level function here (and in other set-up/tear-down logic), and I will update the patch accordingly. As to "why,"--the use of lower-level functions for set-up/tear-down logic in this patch is largely a vestige of my thinking prior to the design discussions we've had since February.

jugglinmike replied:
Turns out we already have a fixture for that!

andreastt commented on webdriver/support/fixtures.py:
```diff
@@ -115,3 +117,83 @@ def inner(path, query="", fragment=""):
+        session.send_session_command("POST",
+                                     "execute/sync",
+                                     dict(script="window.open();", args=[]))
+        with_new = session.send_session_command("GET", "window/handles")
```
Same, but for `session.handles`.

andreastt commented on webdriver/support/fixtures.py:
```diff
@@ -115,3 +117,83 @@ def inner(path, query="", fragment=""):
+            assert result == expected_value, (
+                   "%s dialog was not handled with expected value" % s)
+
+        return assert_dialog_handled
```
There is a lot of indirection and going on in this assert, and many anonymous functions nested within one another. Can you do anything about that? I’m not sure it makes sense to return an assert function from a function to create a dialogue…

jugglinmike replied:
This fixture function does have a lot of functionality. This is because the
task of creating a dialog, checking for the dialog's presence, and validating
its return value all depend on details that can be encapsulated. I thought
doing so would make the tests easier to read/understand. It sounds like it's
too "magical", though, so I've simplified the fixture function and created two
new utility functions (`read_global` and `assert_dialog_handled`) to work with
it.

andreastt replied:
Thanks!

andreastt commented on webdriver/support/fixtures.py:
```diff
@@ -115,3 +117,83 @@ def inner(path, query="", fragment=""):
+
+        assert len(initial) == len(with_new) - 1
+
+        new_handle = (set(with_new) - set(initial)).pop()
```
The order of window handles is not guaranteed, so you need to look at the diff.

jugglinmike replied:
I think that is accounted for by casting the Lists into sets. Although this code is going away, it's the same pattern used in the previously-existing `create_window` fixture function, so please let me know if that's faulty.

andreastt replied:
OK. It might be, because the newest window handle is not guaranteed to be the last one on merged set.

jugglinmike replied:
This set has exactly one element--see the preceding `assert` statement. That
might make using `pop` a little misleading, though. If you'd like, I can update
the existing code to simply reference the first (and only) member via `[0]`.

andreastt replied:
Ah, I see now. This should be fine. Thanks for explaining it to old me.

andreastt commented on webdriver/support/fixtures.py:
```diff
@@ -115,3 +117,83 @@ def inner(path, query="", fragment=""):
         return rv
     return inner
+
+def switch_to_inactive(session):
```
I don’t think this is a particularly obvious name. It tries to do a lot of things. Maybe it would be better to do some of these things in the test.

jugglinmike commented on webdriver/get-title.py:
```diff
@@ -0,0 +1,268 @@
+import pytest
```
Got it.

jugglinmike commented:
@andreastt I think I've addressed all of your feedback. This included a rebase,
so I've made the original version of this branch available elsewhere:

https://github.com/w3c/web-platform-tests/compare/master...bocoup:webdriver-get-title-old

Thanks for the thorough review!

andreastt commented on webdriver/get_title.py:
```diff
@@ -0,0 +1,282 @@
+from support.inline import inline
+
+def read_global(session, name):
+    return session.execute_script("return %s;" % name)
```
I personally don’t feel this function adds any value. `expected_title = session.execute_script("return document.title")` is not substantively longer to write than `expected_title = read_global(session, "document.title")`.

But I guess it’s fine.

andreastt commented on webdriver/support/fixtures.py:
```diff
@@ -142,3 +144,41 @@ def inner(path, query="", fragment=""):
+        assert dialog_type in ("alert", "confirm", "prompt"), (
+               "Invalid dialog type: '%s'" % dialog_type)
+
+        if text == None:
```
s/==/is/

jugglinmike replied:
Got it

andreastt commented on webdriver/support/fixtures.py:
```diff
@@ -142,3 +144,41 @@ def inner(path, query="", fragment=""):
+
+        assert isinstance(text, basestring), "`text` parameter must be a string"
+
+        if result_var == None:
```
s/==/is/

jugglinmike replied:
Got it

andreastt commented on webdriver/support/fixtures.py:
```diff
@@ -142,3 +144,41 @@ def inner(path, query="", fragment=""):
+        if result_var == None:
+            result_var = "__WEBDRIVER"
+
+        assert re.search(r"^[_$a-z$][_$a-z0-9]*$", result_var, re.IGNORECASE)
```
What does this check?

jugglinmike replied:
It ensures that the provided string is a valid JavaScript identifier. No one
likes reading involved regular expressions, though, so I've added an assertion
message to that effect.

andreastt commented on webdriver/support/fixtures.py:
```diff
@@ -142,3 +144,41 @@ def inner(path, query="", fragment=""):
+
+        session.send_session_command("POST",
+                                     "execute/async",
+                                     dict(script=spawn, args=[]))
```
`{"script": spawn, "args": []}`

jugglinmike replied:
Sure

andreastt commented on webdriver/support/fixtures.py:
```diff
@@ -142,3 +144,41 @@ def inner(path, query="", fragment=""):
+            var done = arguments[0];
+            setTimeout(function() {{
+                done();
+            }}, 0);
```
Double `{{` and `}}`?

andreastt replied:
You could also use ES6 arrow functions here.

jugglinmike replied:
Yup. [That's the escape sequences for literal braces in Python's
`string.format` syntax](https://docs.python.org/2/library/string.html):

> Format strings contain “replacement fields” surrounded by curly braces {}.
> Anything that is not contained in braces is considered literal text, which is
> copied unchanged to the output. If you need to include a brace character in
> the literal text, it can be escaped by doubling: {{ and }}.

jugglinmike replied:
Or no function at all--just the `done` reference

andreastt replied:
Ah, I see. Thanks for clarifying.

andreastt replied:
Right, even better!

andreastt commented on webdriver/support/fixtures.py:
```diff
@@ -142,3 +144,41 @@ def inner(path, query="", fragment=""):
+    function to validate that the dialog has been "handled" (either accepted or
+    dismissed) by returning some value."""
+
+    def create_dialog(dialog_type, text=None, result_var=None):
```
Instead of taking a `result_var` to write a custom global, maybe you could generate a variable name and return it from `create_dialog`?

I think I would prefer an API that looked like this (pseudo-code):

```py
my_dialog = create_dialog(session)("prompt", "What do you like most?")
session.alert.set_text("Cheese!")
assert_dialog_accepted(my_dialog, "Cheese!")
```

jugglinmike replied:
Although I do like cheese a lot (it's Havarti for me these days), I'm not so
sure about that API. It seems open to the same criticism that you had for my
original `create_dialog` implementation--the returned value is something (a
global identifier that will hold the result of the expression after dismissal)
very different than what you would expect (a representation of the dialog).
The current implementation is more verbose than both of our ideas, but it's
also the most direct about what is going on.

andreastt replied:
I guess I don’t care enough about this at the moment to start another discussion about this form of API. It’s good enough for now and better than the status quo.

jugglinmike commented:
@andreastt Ready for another round

andreastt reviewed (approved):
r+, but please squash and prepare for integration.

jugglinmike commented:
> r+, but please squash and prepare for integration.

Thanks! I think that https://github.com/w3c/web-platform-tests/pull/5506 is higher priority, though, so I'm going to wait until that lands and rebase. I'll let you know when that happens.

w3c-bots commented:
These tests are now available on [w3c-test.org](https://w3c-test.org/submissions/5275/)

jugglinmike commented:
Hi @andreastt! Now that gh-5506 has landed, I've been able to re-base this
patch to use those improvements. I squashed the original commits. If you'd like
to compare against the original state of this branch, I've made that available
at:

https://github.com/w3c/web-platform-tests/compare/master...bocoup:webdriver-get-title-prebase

What do you think?

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/234927340)

# Firefox (nightly) #
Testing web-platform-tests at revision 79f5a04661632011029056b7a61941b5472ad55c
Using browser at version BuildID 20170522100359; SourceStamp 9851fcb0bf4d855c36729d7de19f0fa5c9f69776
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>9 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5275/webdriver/actions/conftest.py">/webdriver/actions/conftest.py</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | OK      |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5275/webdriver/actions/key.py">/webdriver/actions/key.py</a></summary>


|                                       Subtest                                       | Results | Messages |
|-------------------------------------------------------------------------------------|---------|----------|
|                                                                                     | OK      |          |
| `key.py::test_lone_keyup_sends_no_events`                                           | PASS    |          |
| `key.py::test_single_printable_key_sends_correct_events[0a-KeyA]`                   | PASS    |          |
| `key.py::test_single_printable_key_sends_correct_events[1a-KeyA]`                   | PASS    |          |
| `key.py::test_single_printable_key_sends_correct_events[2"-Quote]`                  | PASS    |          |
| `key.py::test_single_printable_key_sends_correct_events[3,-Comma]`                  | PASS    |          |
| `key.py::test_single_printable_key_sends_correct_events[4value4-]`                  | PASS    |          |
| `key.py::test_single_printable_key_sends_correct_events[5value5-]`                  | PASS    |          |
| `key.py::test_single_printable_key_sends_correct_events[6@-Digit2]`                 | PASS    |          |
| `key.py::test_single_printable_key_sends_correct_events[7value7-]`                  | PASS    |          |
| `key.py::test_single_printable_key_sends_correct_events[8value8-]`                  | PASS    |          |
| `key.py::test_single_emoji_records_correct_key[value0]`                             | PASS    |          |
| `key.py::test_single_emoji_records_correct_key[value1]`                             | PASS    |          |
| `key.py::test_single_modifier_key_sends_correct_events[value0-ShiftRight-Shift]`    | PASS    |          |
| `key.py::test_single_modifier_key_sends_correct_events[value1-OSRight-Meta]`        | PASS    |          |
| `key.py::test_single_modifier_key_sends_correct_events[value2-ControlLeft-Control]` | PASS    |          |
| `key.py::test_single_nonprintable_key_sends_events[value0-Escape-Escape]`           | PASS    |          |
| `key.py::test_single_nonprintable_key_sends_events[value1-ArrowRight-ArrowRight]`   | PASS    |          |
| `key.py::test_sequence_of_keydown_printable_keys_sends_events`                      | PASS    |          |
| `key.py::test_sequence_of_keydown_character_keys`                                   | PASS    |          |
| `key.py::test_backspace_erases_keys`                                                | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5275/webdriver/actions/mouse.py">/webdriver/actions/mouse.py</a></summary>


|                  Subtest                  | Results | Messages |
|-------------------------------------------|---------|----------|
|                                           | OK      |          |
| `mouse.py::test_click_at_coordinates`     | PASS    |          |
| `mouse.py::test_click_element_center`     | PASS    |          |
| `mouse.py::test_click_navigation`         | PASS    |          |
| `mouse.py::test_drag_and_drop[20-0-0]`    | PASS    |          |
| `mouse.py::test_drag_and_drop[20-0-300]`  | PASS    |          |
| `mouse.py::test_drag_and_drop[20-0-800]`  | PASS    |          |
| `mouse.py::test_drag_and_drop[0-15-0]`    | PASS    |          |
| `mouse.py::test_drag_and_drop[0-15-300]`  | PASS    |          |
| `mouse.py::test_drag_and_drop[0-15-800]`  | PASS    |          |
| `mouse.py::test_drag_and_drop[10-15-0]`   | PASS    |          |
| `mouse.py::test_drag_and_drop[10-15-300]` | PASS    |          |
| `mouse.py::test_drag_and_drop[10-15-800]` | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5275/webdriver/actions/sequence.py">/webdriver/actions/sequence.py</a></summary>


|                                 Subtest                                 | Results | Messages |
|-------------------------------------------------------------------------|---------|----------|
|                                                                         | OK      |          |
| `sequence.py::test_no_actions_send_no_events`                           | PASS    |          |
| `sequence.py::test_release_char_sequence_sends_keyup_events_in_reverse` | PASS    |          |
| `sequence.py::test_release_no_actions_sends_no_events`                  | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5275/webdriver/actions/special_keys.py">/webdriver/actions/special_keys.py</a></summary>


|                                        Subtest                                         | Results | Messages |
|----------------------------------------------------------------------------------------|---------|----------|
|                                                                                        | OK      |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[NUMPAD9-expected0]`         | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[RETURN-expected1]`          | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[HELP-expected2]`            | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[SHIFT-expected3]`           | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[R_ARROWRIGHT-expected4]`    | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[ESCAPE-expected5]`          | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[PAGE_UP-expected6]`         | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[R_PAGEUP-expected7]`        | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[UP-expected8]`              | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[DOWN-expected9]`            | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[F12-expected10]`            | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[META-expected11]`           | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[BACKSPACE-expected12]`      | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[MULTIPLY-expected13]`       | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[HOME-expected14]`           | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[NULL-expected15]`           | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[SUBTRACT-expected16]`       | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[CONTROL-expected17]`        | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[INSERT-expected18]`         | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[R_META-expected19]`         | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[SEMICOLON-expected20]`      | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[SPACE-expected21]`          | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[NUMPAD4-expected22]`        | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[RIGHT-expected23]`          | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[TAB-expected24]`            | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[R_ALT-expected25]`          | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[NUMPAD0-expected26]`        | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[DECIMAL-expected27]`        | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[LEFT-expected28]`           | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[R_DELETE-expected29]`       | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[PAGE_DOWN-expected30]`      | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[PAUSE-expected31]`          | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[END-expected32]`            | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[DIVIDE-expected33]`         | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[R_ARROWUP-expected34]`      | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[NUMPAD3-expected35]`        | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[CLEAR-expected36]`          | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[R_ARROWLEFT-expected37]`    | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[EQUALS-expected38]`         | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[R_PAGEDOWN-expected39]`     | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[ADD-expected40]`            | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[NUMPAD1-expected41]`        | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[R_INSERT-expected42]`       | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[ENTER-expected43]`          | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[CANCEL-expected44]`         | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[NUMPAD6-expected45]`        | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[F10-expected46]`            | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[F11-expected47]`            | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[R_END-expected48]`          | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[NUMPAD7-expected49]`        | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[NUMPAD2-expected50]`        | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[F1-expected51]`             | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[F2-expected52]`             | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[F3-expected53]`             | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[F4-expected54]`             | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[F5-expected55]`             | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[F6-expected56]`             | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[F7-expected57]`             | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[F8-expected58]`             | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[F9-expected59]`             | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[NUMPAD8-expected60]`        | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[NUMPAD5-expected61]`        | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[R_CONTROL-expected62]`      | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[R_HOME-expected63]`         | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[ZENKAKUHANKAKU-expected64]` | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[R_SHIFT-expected65]`        | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[SEPARATOR-expected66]`      | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[ALT-expected67]`            | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[R_ARROWDOWN-expected68]`    | PASS    |          |
| `special_keys.py::test_webdriver_special_key_sends_keydown[DELETE-expected69]`         | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5275/webdriver/conftest.py">/webdriver/conftest.py</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | OK      |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5275/webdriver/contexts.py">/webdriver/contexts.py</a></summary>


|                  Subtest                  | Results | Messages |
|-------------------------------------------|---------|----------|
|                                           | OK      |          |
| `contexts.py::test_window_size_types`     | PASS    |          |
| `contexts.py::test_window_resize`         | PASS    |          |
| `contexts.py::test_window_position_types` | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5275/webdriver/get_title.py">/webdriver/get_title.py</a></summary>


|                        Subtest                         | Results | Messages |
|--------------------------------------------------------|---------|----------|
|                                                        | OK      |          |
| `get_title.py::test_title_from_closed_context`         | PASS    |          |
| `get_title.py::test_title_handle_prompt_dismiss`       | FAIL    |          |
| `get_title.py::test_title_handle_prompt_accept`        | FAIL    |          |
| `get_title.py::test_title_handle_prompt_missing_value` | FAIL    |          |
| `get_title.py::test_title_from_top_context`            | PASS    |          |
| `get_title.py::test_title_with_duplicate_element`      | PASS    |          |
| `get_title.py::test_title_without_element`             | PASS    |          |
| `get_title.py::test_title_after_modification`          | PASS    |          |
| `get_title.py::test_title_strip_and_collapse`          | PASS    |          |
| `get_title.py::test_title_from_frame`                  | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5275/webdriver/navigation.py">/webdriver/navigation.py</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | TIMEOUT |          |

</details>
</details>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/234927341)

# Sauce (safari) #
Testing web-platform-tests at revision 79f5a04661632011029056b7a61941b5472ad55c
Using browser at version 10.0
Starting 10 test iterations
No tests run.

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/234927347)

# Chrome (unstable) #
Testing web-platform-tests at revision 79f5a04661632011029056b7a61941b5472ad55c
Using browser at version 60.0.3100.0 dev
Starting 10 test iterations
No tests run.

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/234927348)

# Sauce (MicrosoftEdge) #
Testing web-platform-tests at revision 79f5a04661632011029056b7a61941b5472ad55c
Using browser at version 14.14393
Starting 10 test iterations
No tests run.

andreastt commented:
@jugglinmike 👍  That looks fine, please go ahead!

jugglinmike committed:
fixup! [webdriver] Add tests for "Get Title" command
